### PR TITLE
sql: allow filtering subsources by upstream schema

### PR DESCRIPTION
**Pull request by sploiselle**
To make PG sources more ergonomic, I suggest letting users scope the tables they ingest by schema, not just "all" or a "specific list," e.g. I imagine a lot of people want to scope this to `public` and not ingest every table they've ever created. This also takes a step toward handling some of the complexity of helping users map their upstream namespaces to their subsources.

### Motivation

This PR adds a feature that has not yet been specified. [design doc](https://github.com/MaterializeInc/materialize/pull/18338)

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Let users include only tables from particular schemas when creating PostgreSQL sources using the new `FOR SCHEMAS(...)` clause in `CREATE SOURCE`.

### CODE REVIEW

**Review by jkosh44 — APPROVED** [review]
surfaces changes LGTM, but you probably want to wait for a storage approval too.

File: src/sql/src/pure.rs
Comment (petrosagg) [inline src/sql/src/pure.rs:356]: I'm not sure if we should error based off of the list of publication tables. Given the list of published tables we can't really tell if a schema is published in full or partially and a list of publications without any table of a given schema can very well be interpreted as an empty schema, which results in a no-op.

I would suggest that we don't error here and instead check if the schema exists in the database by inspecting it during purification. If it doesn't we can error there and here we just treat valid schemas with no published tables as no-ops.

**Review by petrosagg — APPROVED** [review]
Looks good but I would change the error behavior for schemas that have no tables published on the basis that this could happen with automated scripts

File: doc/user/content/sql/create-source/postgres.md
Comment (morsapaes) [reply doc/user/content/sql/create-source/postgres.md:35]: Using the RR diagram sounds good. There's just a mix of approaches right now (see [this example](https://materialize.com/docs/sql/create-connection/#kafka-options)), so we should guarantee that the end state is consistent. @MaterializeInc/docs

(Ugh, there's something funny going on with the connections page. There was a recent PR that refactored it, will have a look.)

**Pull request merged**
